### ISSUE TITLE
Can't get table metadata for tables with '(' in their names

### ISSUE THREAD

**Issue by trombipeti** [yiisoft/yii2#18001]
Disclaimer: I know it is a horrible idea to have table names like this, but I have to work with a DB that already contains some and I'd like for my code to work nonetheless.

### What steps will reproduce the problem?
Create a table in the database of the app with parenthesis in the table name, so like:
``CREATE TABLE `example_table(0)` (
 `ID` int(11) NOT NULL AUTO_INCREMENT,
 PRIMARY KEY (`ID`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8``
Try to run this line or any code that involves querying the tables schema.
`Yii::$app->db->getTableSchema("example_table(0)", true);`
### What is the expected result?
I expect to get the table metadata returned, as it is with any other tables without the parenthesis.

### What do you get instead?
`SQLSTATE[42000]: Syntax error or access violation: 1064 You have an error in your SQL syntax; check the manual that corresponds to your MariaDB server version for the right syntax to use near '(0)' at line 1
The SQL being executed was: SHOW FULL COLUMNS FROM example_table(0)`
As you can see, the table name isn't quoted correctly. Looking at the code makes me believe that this issue is present independent of the DB driver.
https://github.com/yiisoft/yii2/blob/dd657f8d64dc24b842f5b65a0e8748cd78bf6554/framework/db/Schema.php#L481-L483
Moreover, it doesn't help if I quote the table name myself, because the quoting characters are removed when building the query. I _think_ this only happens when the MySQL driver is used.

### Additional info

| Q                | A
| ---------------- | ---
| Yii version      | 2.0.34
| PHP version      | 7.4.5
| Operating system | Linux

**Comment by samdark** [comment]
Yes, it could be considered a bug. A very rare one. Fixing it is not a priority but we'll accept a pull request with tests and a fix.

**Comment by trombipeti** [comment]
Ok, thanks!
Any background on why it is like this? The '{{' part I get but I couldn't figure out why having a '(' in the table name should prevent it from being quoted.

**Comment by samdark** [comment]
> Any background on why it is like this?

We haven't thought of such names.

> but I couldn't figure out why having a '(' in the table name should prevent it from being quoted

I think that's about SQL for create/alter table or column. `(` is used there to specify size of the column.

**Comment by mbunyan** [comment]
The sql statement needs to include backticks(`) which would cover the majority of MySQL installations. The exception is where ANSI_QUOTES is set to true using: SET sql_mode='ANSI_QUOTES'; (rare IMHO)

The brackets are allowed characters but MUST be quoted as they occur in Unicode BMP

[The reference in MySQL 5.7 is same as 8](https://dev.mysql.com/doc/refman/5.7/en/identifiers.html) 

Permitted characters in unquoted identifiers: ASCII: [0-9,a-z,A-Z$_] (basic Latin letters, digits 0-9, dollar, underscore), Extended: U+0080 .. U+FFFF

Permitted characters in quoted identifiers include the full Unicode Basic Multilingual Plane (BMP), except U+0000: ASCII: U+0001 .. U+007F, Extended: U+0080 .. U+FFFF

ASCII NUL (U+0000) and supplementary characters (U+10000 and higher) are not permitted in quoted or unquoted identifiers. Identifiers may begin with a digit but unless quoted may not consist solely of digits. Database, table, and column names cannot end with space characters.

The identifier quote character is the backtick (`).

**Comment by trombipeti** [comment]
> I think that's about SQL for create/alter table or column. `(` is used there to specify size of the column.

But that doesn't have to do anything with the table name, does it? I believe it should be fine then to just remove the check for '(' in quoteTableName.
The only thing I fear that there's someone out there depending on this behavior.

@qiangxue, can you maybe tell why this was introduced like this? I think it was this commit:
https://github.com/yiisoft/yii2/commit/fe90d4dda05debec54cd55d597e01c47bec1b980

I went forward and tried table names with '{{' in them, they are of course also not quoted, but those {{ get stripped along the way - they are there to support table prefixes. So I think that's fine, although it might deserve a mention in the documentation that Yii2 can't load table schemas for such tables.

**Comment by floor12** [comment]
The same problem of course presents with postresql. 
I also dont understand why `yii\db\Schema::quoteTableName()` has return in case of `(`. Maybe we could remove this part...

**Comment by samdark** [comment]
Maybe. First step would be removing it, making a pull request and watching for test results. If it would fail or not. @floor12 do you have some time do it?

**Comment by floor12** [comment]
@samdark yes, I will do it. If all tests will pass, maybe we need to add cases with `(` `)`  to  the tests. I will check.

**Comment by floor12** [comment]
So, I updated method `yii\db\Schema::quoteTableName()` 
from
`if (strpos($name, '(') !== false || strpos($name, '{{') !== false) {`
to
`if (strpos($name, '{{') !== false) {`.

After this action a found this assertion in `yiiunit\framework\db\ConnectionTest`:
`$this->assertEquals('(table)', $connection->quoteTableName('(table)'));`

So, there is special test for this case and of course it went to fail.
We need to understand for what case it was done in general. Maybe some databases doesn't support brackets in the table names... Or something else. 

After updating test to 
`$this->assertEquals('`(table)`', $connection->quoteTableName('(table)'));` all run well in tests and real Postgres creates tables with brackets successful.

**Comment by rob006** [comment]
This condition may be for cases where expression is used instead of table name (`SELECT * FROM (SELECT * FROM some_table)`). It does not look like good place for this check (if you use `quoteTableName()` the you should pass table name, not expression), so I would double check where `quoteTableName()` is used internally and make sure that expression are not passed to this method.

**Comment by trombipeti** [comment]
To me it looks as if the Connection/Schema isn't a 100% general class, but supports some internal Yii2 stuff needed by e.g. ActiveRecord. That is why I was afraid to just do what @floor12 tried, because it could break more things than you'd think first.

**Comment by floor12** [comment]
To fix this issue we can check if the bracket `(` exists and if it is the first letter in table name, then return it without quote. It will not quote the expressions witch starts with `(` but quote table names with brackets like `table_name(1)` etc. 

To do this trick we just need to replace `false` to `0` like this:
```php
if (strpos($name, '(') !== 0 || strpos($name, '{{') !== false) {
```

Or we can add additional check to latest bracket and separate it from `{{` checking like this:

```php
if (strpos($name, '(') === 0 && strpos($name, ')') === strlen($name) - 1) {
    return $name;
}
if (strpos($name, '{{') !== false) {
    return $name;
}
```

**Comment by samdark** [comment]
@floor12 worth trying in the same `Fix#18001-quote-table-name` branch.
